homeassistant/components/fronius/sensor.py

```diff
@@ -0,0 +1,184 @@
+"""Support for Fronius devices."""
+import logging
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (CONF_RESOURCE, CONF_SENSOR_TYPE, CONF_DEVICE,
+                                 CONF_MONITORED_CONDITIONS)
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_SCOPE = 'scope'
+
+TYPE_INVERTER = 'inverter'
+TYPE_STORAGE = 'storage'
+TYPE_METER = 'meter'
+TYPE_POWER_FLOW = 'power_flow'
+SCOPE_DEVICE = 'device'
+SCOPE_SYSTEM = 'system'
+
+DEFAULT_SCOPE = SCOPE_DEVICE
+DEFAULT_DEVICE = 0
+
+SENSOR_TYPES = [TYPE_INVERTER, TYPE_STORAGE, TYPE_METER, TYPE_POWER_FLOW]
+SCOPE_TYPES = [SCOPE_DEVICE, SCOPE_SYSTEM]
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_RESOURCE):
+    cv.url,
+    vol.Required(CONF_MONITORED_CONDITIONS):
+    vol.All(
+        cv.ensure_list,
+        [{
+            vol.Required(CONF_SENSOR_TYPE): vol.In(SENSOR_TYPES),
+            vol.Optional(CONF_SCOPE, default=DEFAULT_SCOPE):
+                vol.In(SCOPE_TYPES),
+            vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE):
+                vol.All(vol.Coerce(int), vol.Range(min=0))
+        }]
+    )
+})
+
+
+async def async_setup_platform(hass,
+                               config,
+                               async_add_devices,
+                               discovery_info=None):
+    """Set up of Fronius platform."""
+    from pyfronius import Fronius
+
+    session = async_get_clientsession(hass)
+    fronius = Fronius(session, config[CONF_RESOURCE])
+
+    sensors = []
+    for condition in config[CONF_MONITORED_CONDITIONS]:
+
+        device = condition.get(CONF_DEVICE)
+        if device == 0:
+            if condition[CONF_SENSOR_TYPE] == TYPE_INVERTER:
+                device = 1
```

Comment:
All config modifications should be made in the config schema. Please make a custom validator that checks the condition and makes this modification before returning the config part.

Here's an example of a custom validator that modifies the config:
https://github.com/home-assistant/home-assistant/blob/a89c8eeabe6b9e847c33bce2ef3987b3caf387ad/homeassistant/components/manual/alarm_control_panel.py#L46-L58

Reply:
I hope the way I implemented it now is the way you intended. It was tested to work locally at least.